pkg/log/zap/kube_helpers.go

```diff
@@ -57,14 +62,21 @@ func (w namespacedNameWrapper) MarshalLogObject(enc zapcore.ObjectEncoder) error
 
 // kubeObjectWrapper is a zapcore.ObjectMarshaler for Kubernetes objects.
 type kubeObjectWrapper struct {
-	obj runtime.Object
+	obj    runtime.Object
+	scheme *runtime.Scheme
 }
 
 // MarshalLogObject implements zapcore.ObjectMarshaler
 func (w kubeObjectWrapper) MarshalLogObject(enc zapcore.ObjectEncoder) error {
-	// TODO(directxman12): log kind and apiversion if not set explicitly (common case)
-	// -- needs an a scheme to convert to the GVK.
 	gvk := w.obj.GetObjectKind().GroupVersionKind()
+	if gvk.Version == "" && w.scheme != nil {
+		var err error
+		gvk, err = apiutil.GVKForObject(w.obj, w.scheme)
+		if err != nil {
+			return fmt.Errorf("cannot deduce GVK for object: %v", w.obj)
```

Comment:
is it more useful to do this, or to just continue on with an error or some sort of sentinel value (e.g. `__unknown` or somesuch)?

I think the latter, but I'm willing to hear counter-arguments.

Reply:
Since this is logging, I'd say it's more useful giving any information over crashing with an error. In the common use-case, the schema should have all types registered. 

I'd argue the most useful value would be 

```go
gvk.Kind = fmt.Sprintf("%T", w.obj)
gvk.Version = "__VersionInternal" // IIRC there's some constant for this when I worker with extension API server, though I'm not 100% in which package it is. 
gvk.Group= "__unknown; object not registered in logger schema"
```   

What I'd like presenting to the user is both some useful information (the object's type), and how can he improve the logging information (by adding the type in the schema). 

What do you think?

Reply:
that seems reasonable, although I'd avoid the spaces in Group and just go with `Kind: fmt.Sprintf("%T", w.obj), gvk.Version: "__unknown", gvk.Group: "__unknown"}`

Reply:
done